File: apps/php/src/Domain/Type/FizzBuzzTypeName.php
<?php

declare(strict_types=1);

namespace App\Domain\Type;

enum FizzBuzzTypeName: int
{
    case Standard = 1;
    case NumberOnly = 2;
    case FizzBuzzOnly = 3;

    public function createType(): FizzBuzzType
    {
        return match ($this) {
            self::Standard => new FizzBuzzType01(),
            self::NumberOnly => new FizzBuzzType02(),
            self::FizzBuzzOnly => new FizzBuzzType03(),
        };
    }

    public function label(): string
    {
        return match ($this) {
            self::Standard => '通常',
            self::NumberOnly => '数値のみ',
            self::FizzBuzzOnly => 'FizzBuzzのみ',
        };
    }
}

File: apps/rust/src/domain/types/fizz_buzz_type_name.rs
use super::{FizzBuzzType, FizzBuzzType01, FizzBuzzType02, FizzBuzzType03};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum FizzBuzzTypeName {
    Standard = 1,
    NumberOnly = 2,
    FizzBuzzOnly = 3,
}

impl FizzBuzzTypeName {
    pub fn from_number(n: i32) -> Result<Self, String> {
        match n {
            1 => Ok(Self::Standard),
            2 => Ok(Self::NumberOnly),
            3 => Ok(Self::FizzBuzzOnly),
            _ => Err(format!("タイプ{}は見つかりません", n)),
        }
    }

    pub fn create_type(&self) -> Box<dyn FizzBuzzType> {
        match self {
            Self::Standard => Box::new(FizzBuzzType01),
            Self::NumberOnly => Box::new(FizzBuzzType02),
            Self::FizzBuzzOnly => Box::new(FizzBuzzType03),
        }
    }

    pub fn label(&self) -> &str {
        match self {
            Self::Standard => "通常",
            Self::NumberOnly => "数値のみ",
            Self::FizzBuzzOnly => "FizzBuzzのみ",
        }
    }
}
